interfaces/prompting/patterns/render.go

```diff
@@ -0,0 +1,376 @@
+	var buf bytes.Buffer
+	var moreRemain bool
+
+	c := n.Config()
+	length := 0
+	lengthUnchanged := 0
+
+	for i := 0; ; i++ {
+		// TODO: change how buffer is handled, so that we only need to re-render
+		// the part that was invalidated by the old config.
+		buf.Truncate(lengthUnchanged)
+		n.Render(&buf, c, lengthUnchanged)
+		observe(i, &buf)
+		length, lengthUnchanged, moreRemain = c.NextEx(n)
+		if !moreRemain {
+			break
+		}
+		buf.Grow(length)
+	}
+}
+
+// Literal is a render node with a literal string.
+type Literal string
+
+func (n Literal) Render(buf *bytes.Buffer, conf RenderConfig, alreadyWritten int) int {
+	if alreadyWritten > 0 {
+		return alreadyWritten - len(n)
+	}
+	buf.WriteString(string(n))
+	return 0
+}
+
+func (n Literal) NumVariants() int {
+	return 1
+}
+
+func (n Literal) Config() RenderConfig {
+	return literalConfig{}
+}
+
+func (n Literal) nodeEqual(other RenderNode) bool {
+	if other, ok := other.(Literal); ok {
+		return n == other
+	}
+
+	return false
+}
+
+type literalConfig struct{}
+
+func (literalConfig) NextEx(n RenderNode) (length, lengthUnchanged int, moreRemain bool) {
+	literal := n.(Literal)
+	return len(literal), 0, false
+}
+
+func (literalConfig) Length(n RenderNode) int {
+	literal := n.(Literal)
+	return len(literal)
+}
+
+// Seq is sequence of consecutive render nodes.
+type Seq []RenderNode
+
+func (n Seq) Render(buf *bytes.Buffer, conf RenderConfig, alreadyWritten int) int {
+	c := conf.(seqConfig)
+
+	for i := range n {
+		alreadyWritten = n[i].Render(buf, c[i], alreadyWritten)
+	}
+
+	return alreadyWritten
+}
+
+func (n Seq) NumVariants() int {
+	num := 1
+
+	for i := range n {
+		num *= n[i].NumVariants()
+	}
+
+	return num
+}
+
+func (n Seq) Config() RenderConfig {
+	if len(n) == 0 {
+		return seqConfig(nil)
+	}
+
+	c := make(seqConfig, len(n))
+
+	for i := range n {
+		c[i] = n[i].Config()
+	}
+
+	return c
+}
+
+func (n Seq) nodeEqual(other RenderNode) bool {
+	if other, ok := other.(Seq); ok {
+		if len(other) != len(n) {
+			return false
+		}
+
+		for i := range n {
+			if !n[i].nodeEqual(other[i]) {
+				return false
+			}
+		}
+	}
+
+	return false
+}
+
+func (seq Seq) reduceStrength() RenderNode {
+	switch len(seq) {
+	case 0:
+		return Literal("")
+	case 1:
+		return seq[0]
+	default:
+		return seq
+	}
+}
+
+func (seq Seq) optimize() Seq {
+	var b strings.Builder
+
+	var newSeq Seq
+
+	for _, item := range seq {
+		if v, ok := item.(Literal); ok {
+			if v == "" {
+				continue
+			}
+
+			b.WriteString(string(v))
+		} else {
+			if b.Len() > 0 {
+				newSeq = append(newSeq, Literal(b.String()))
+				b.Reset()
+			}
+
+			newSeq = append(newSeq, item)
+		}
+	}
+
+	if b.Len() > 0 {
+		newSeq = append(newSeq, Literal(b.String()))
+		b.Reset()
+	}
+
+	return newSeq
+}
+
+// seqConfig is the configuration for a seqeunce of render nodes.
+//
+// Each render node has a corresponding configuration at the same index.
+type seqConfig []RenderConfig
```

Comment:
I see it's convenient type wise not to have to carry the node here, but then given the strange API of NextEx it's not a clear win confusion wise

Reply:
I agree, it is a it strange. Curious what @zyga thinks, could we just embed a pointer to the renderNode in each `variantState` (formerly `RenderConfig`) implementation?

Reply:
I've done this, see comment above.